sharktank/sharktank/layers/ffn_moe_block.py

```diff
@@ -74,7 +77,105 @@ def forward(
+        theta: Theta,
+        rms_epsilon: float | None = None,
+        is_gated: bool = True,
+        activation_fn: Callable[[torch.Tensor], torch.Tensor] = F.silu,
+        activation_dtype: Optional[torch.dtype] = None,
+        fake_quant: bool = False,
+    ):
+        super().__init__(theta)
+        self.num_experts = theta("ffn_gate_exps", "weight").shape[0]
+        ffn_theta = Theta(
+            {
+                "ffn_gate": theta("ffn_gate_exps").tree,
+                "ffn_up": theta("ffn_up_exps").tree,
+                "ffn_down": theta("ffn_down_exps").tree,
+            }
+        )
+        self.ffn = FFN(
+            ffn_theta,
+            rms_epsilon=rms_epsilon,
+            is_gated=is_gated,
+            activation_fn=activation_fn,
+            activation_dtype=activation_dtype,
+            fake_quant=fake_quant,
+            add_residual=False,
+        )
+
+    def forward(
+        self,
+        h: torch.Tensor,
+        top_experts_index: torch.Tensor,
+        expert_gate: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        h:
+            Input tokens.
+            Shape of (batch_size * sequence_length, input_feature_dim).
+        top_experts_index:
+            indexes of selected top experts for each token.
+            Shape of (batch_size * sequence_length, num_top_experts).
+            Value range is [0, num_of_experts)
+        expert_gate:
+            router weights of selected top experts for each token.
+            Shape of (batch_size * sequence_length, num_top_experts).
+
+        Returns:
+            Tensor of shape (batch_size * sequence_length, expert_feature_dim)
+        """
+        num_tokens, input_feature_dim = h.shape
+
+        # (self.num_experts, num_tokens)
+        router_scores = (
+            torch.full([num_tokens, self.num_experts], 0, dtype=h.dtype)
+            .scatter_(1, top_experts_index, expert_gate)
+            .transpose(0, 1)
+        )
+
+        # (self.num_experts, num_tokens)
+        router_indices = (
+            torch.arange(num_tokens, device=h.device)
+            .view(1, -1)
+            .expand(self.num_experts, -1)
+        )
+        # (self.num_experts * num_tokens, input_feature_dim)
+        router_indices = router_indices.reshape(-1, 1).expand(-1, input_feature_dim)
+
+        # (self.num_experts * num_tokens, input_feature_dim)
+        routed_in = ops.gather(
+            input=h,
+            dim=0,
+            index=router_indices,
+        )
+        routed_in = routed_in * (router_scores > 0).reshape(-1, 1)
+        routed_in = routed_in.view(self.num_experts, num_tokens, input_feature_dim)
+
+        # (self.num_experts * num_tokens, input_feature_dim)
+        routed_out = self.ffn(routed_in).view(
+            self.num_experts * num_tokens, input_feature_dim
+        )
+        routed_out = routed_out * router_scores.reshape(-1, 1)
+
+        # (num_tokens, input_feature_dim)
+        return ops.zeros_like(h).scatter_add(
```

Comment:
Against about specialized operations. `scatter_add` feels weird even because if the input is zero this would just be scatter anyway.

Reply:
The mapping is many-to-one. Meaning that multiple addition may occur for the same destination location.